### ISSUE TITLE
GeomVertexData is duplicated for each copy of a hardware-animated model

### ISSUE THREAD

**Issue by rdb** [panda3d/panda3d#421]
When loading multiple independently-animated copies of a model, Panda3D duplicates the GeomVertexArrayData objects for each copy.  This is because GeomVertexData (which lives under the Geom) is storing the TransformTable that reference the joints on the Character node. When you load the model again, it copies the scene graph—including all the Character nodes—which then requires duplicating the all the TransformTable objects because they reference this Character, meaning there need to be as many different GeomVertexData objects as there are copies of the actor.

So, even if hardware skinning is used, Panda nonetheless ends up creating and uploading a separate VBO for every instance.  This is not necessary because the vertex data is still the same for all copies, so we should be able to benefit from the COW mechanism here.

The most obvious way to fix this would be to move the TransformTable to the Geom level; this will still cause 1000 separate Geoms, but at least the underlying vertex and index buffers won't be duplicated except if the animation happens on the CPU.

The caveat however is that there is no way to do this without breaking the API.  Code that procedurally creates animated models or access animated vertex data of existing models will need to be changed to access these things on the Geom instead of on the GeomVertexData.

_Originally posted by @rdb in https://github.com/panda3d/panda3d/issues/292#issuecomment-382496390_

**Comment by Moguri** [comment]
Could we allow storing the TransformTable in either the GeomVertexData or the Geom with the one on the Geom taking precedence if it is there? From there we could deprecate TransformTables in the GeomVertexData.

**Comment by rdb** [comment]
Yeah, hmm.  To support the most common use cases for creating geometry, we could still allow `set_transform_table` on the GeomVertexData when it's newly created, but have it transfer to the `Geom` as soon as it is assigned to the Geom.

For the use cases of reading the transform blend table from a loaded model, I don't see how we can keep supporting it without wiping away the benefits of this change; as soon as we promise that `get_transform_blend_table()` (and `get_slider_table()`, etc.) returns something meaningful, we become obligated to duplicate the GeomVertexData whenever the Character is copied.

**Comment by Moguri** [comment]
I was thinking things like GeomVertexData::get_transform_blend_table() would only return a table if one was written to the GeomVertexData, but I did not think of the case where a user may be trying to read the table without having written/generated it.

**Comment by Moguri** [comment]
Could the behavior be controlled by a PRC variable? We could default the behavior to the legacy behavior and mark it as deprecated. Then in another version we could default to the new behavior. Eventually we could get rid of the PRC variable.

Do you have any numbers to characterize the performance difference?

**Comment by theclashingfritz** [comment]
I do think it would be a major difference, There is instances in Toontown private servers where loads of models are loaded at once, Mostly the same one. This would reduce the work required for these models drastically. I don't have numbers for this yet. But if i get some i'll be sure to post them here.
